test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java

```diff
@@ -20,60 +20,106 @@
             final int mod = field.getModifiers();
-            if (Modifier.isStatic(mod) && Modifier.isFinal(mod) && Modifier.isPublic(mod)) {
-                if (field.getType() == Version.class) {
-                    final int id;
-                    try {
-                        id = ((Version) field.get(null)).id;
-                    } catch (final IllegalAccessException e) {
-                        throw new RuntimeException(e);
-                    }
-                    assert field.getName().matches("(V(_\\d+)+(_(alpha|beta|rc)\\d+)?(_UNRELEASED)?|CURRENT)") : field.getName();
-                    // note that below we remove CURRENT and add it to released; we do it this way because there are two constants that
-                    // correspond to CURRENT, CURRENT itself and the actual version that CURRENT points to
-                    if (field.getName().equals("CURRENT") || field.getName().endsWith("UNRELEASED")) {
-                        unreleasedIdsSet.add(id);
-                    } else {
-                        releasedIdsSet.add(id);
-                    }
-                }
+            if (false == Modifier.isStatic(mod) && Modifier.isFinal(mod) && Modifier.isPublic(mod)) {
+                continue;
+            }
+            if (field.getType() != Version.class) {
+                continue;
+            }
+            assert field.getName().matches("(V(_\\d+)+(_(alpha|beta|rc)\\d+)?|CURRENT)") : field.getName();
+            if (false == VERSION_NAME.matcher(field.getName()).matches()) {
+                continue;
+            }
+            try {
+                versions.add(((Version) field.get(null)));
+            } catch (final IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        Collections.sort(versions);
+        assert versions.get(versions.size() - 1).equals(current) : "The highest version must be the current one "
+            + "but was [" + versions.get(versions.size() - 1) + "] and current was [" + current + "]";
+
+        if (current.revision != 0) {
+            /* If we are in a patch release there should be no unreleased version constants.
+             * If there are, gradle will yell about it. */
+            return new Tuple<>(unmodifiableList(versions), emptyList());
+        }
+        /* If we are on a patch release then we know that at least the version before the
+         * current one is unreleased. If it is released then gradle would be complaining. */
+
+        /* Technically we don't support backwards compatiblity for alphas, betas,
+         * and rcs. But the testing infrastructure requires that we act as though we
+         * do. This is a difference between the gradle and Java logic but should be
+         * fairly safe as it is errs on us being more compatible rather than less....
+         * Anyway, the upshot is that we never declare alphas as unreleased, no
+         * matter where they are in the list */
```

Comment:
I think this is leftovers. Something like this PR will help us level set how we deal with these and then we can clean them up from there. Like, we can remove this funny exception for alphas in a followup, tweaking all the tests that need it.